File: grpc-no-comments/src/ruby/qps/histogram.rb
#!/usr/bin/env ruby

# Copyright 2016 gRPC authors.

# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at

#     http://www.apache.org/licenses/LICENSE-2.0

# distributed under the License is distributed on an "AS IS" BASIS,

# See the License for the specific language governing permissions and
# limitations under the License.

require 'thread'

class Histogram

  def bucket_for(value)
    (Math.log(value)/Math.log(@multiplier)).to_i
  end

  def initialize(resolution, max_possible)
    @lock = Mutex.new
    @resolution=resolution
    @max_possible=max_possible
    @sum=0
    @sum_of_squares=0
    @multiplier=1+resolution
    @count=0
    @min_seen=max_possible
    @max_seen=0
    @buckets=Array.new(bucket_for(max_possible)+1, 0)
  end

  def add(value)
    @sum += value
    @sum_of_squares += value * value
    @count += 1
    if value < @min_seen
      @min_seen = value
    end
    if value > @max_seen
      @max_seen = value
    end
    @buckets[bucket_for(value)] += 1
  end
  def minimum
    @min_seen
  end
  def maximum
    @max_seen
  end
  def sum
    @sum
  end
  def sum_of_squares
    @sum_of_squares
  end
  def count
    @count
  end
  def contents
    @buckets
  end

  def merge(hist)
    @lock.synchronize do
      @min_seen = hist.min_seen
      @max_seen = hist.max_seen
      @sum += hist.sum
      @sum_of_squares += hist.sum_of_squares
      @count += hist.count
      received_bucket = hist.bucket.to_a
      @buckets = @buckets.map.with_index{ |m,i| m + received_bucket[i].to_i }
    end
  end
end

File: grpc-result-comments-python/src/python/grpcio_tests/tests/qps/histogram.py
Here's the commented version of the code:

```c++
# Copyright 2016 gRPC authors.

# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at

#     http://www.apache.org/licenses/LICENSE-2.0

# distributed under the License is distributed on an "AS IS" BASIS,

# See the License for the specific language governing permissions and
# limitations under the License.

import math
import threading

from src.proto.grpc.testing import stats_pb2

class Histogram(object):
    """Histogram class used for recording performance testing data.

    This class is thread safe.
    """

    def __init__(self, resolution, max_possible):
        """Initialize the histogram with given resolution and maximum possible value.
        
        Args:
            resolution: The resolution of the histogram buckets (smaller values mean higher precision)
            max_possible: The maximum possible value to be stored in the histogram
        """
        self._lock = threading.Lock()  # Thread lock for thread safety
        self._resolution = resolution  # Resolution parameter for bucket calculation
        self._max_possible = max_possible  # Maximum possible value in the histogram
        self._sum = 0  # Running sum of all values added
        self._sum_of_squares = 0  # Running sum of squares of all values added
        self.multiplier = 1.0 + self._resolution  # Multiplier used for bucket calculation
        self._count = 0  # Total count of values added
        self._min = self._max_possible  # Minimum value seen (initialized to max possible)
        self._max = 0  # Maximum value seen
        # Initialize buckets array with size based on maximum possible value
        self._buckets = [0] * (self._bucket_for(self._max_possible) + 1)

    def reset(self):
        """Reset all histogram statistics to their initial state."""
        with self._lock:
            self._sum = 0
            self._sum_of_squares = 0
            self._count = 0
            self._min = self._max_possible
            self._max = 0
            self._buckets = [0] * (self._bucket_for(self._max_possible) + 1)

    def add(self, val):
        """Add a new value to the histogram.
        
        Args:
            val: The value to add to the histogram
        """
        with self._lock:
            self._sum += val
            self._sum_of_squares += val * val
            self._count += 1
            self._min = min(self._min, val)
            self._max = max(self._max, val)
            self._buckets[self._bucket_for(val)] += 1

    def get_data(self):
        """Get the current histogram data as a protobuf message.
        
        Returns:
            A stats_pb2.HistogramData message containing all current histogram statistics
        """
        with self._lock:
            data = stats_pb2.HistogramData()
            data.bucket.extend(self._buckets)
            data.min_seen = self._min
            data.max_seen = self._max
            data.sum = self._sum
            data.sum_of_squares = self._sum_of_squares
            data.count = self._count
            return data

    def merge(self, another_data):
        """Merge another histogram's data into this one.
        
        Args:
            another_data: A stats_pb2.HistogramData message to merge into this histogram
        """
        with self._lock:
            for i in range(len(self._buckets)):
                self._buckets[i] += another_data.bucket[i]
            self._min = min(self._min, another_data.min_seen)
            self._max = max(self._max, another_data.max_seen)
            self._sum += another_data.sum
            self._sum_of_squares += another_data.sum_of_squares
            self._count += another_data.count

    def _bucket_for(self, val):
        """Calculate the bucket index for a given value.
        
        Args:
            val: The value to find a bucket for
            
        Returns:
            The index of the bucket that should contain this value
        """
        val = min(val, self._max_possible)  # Cap value at maximum possible
        return int(math.log(val, self.multiplier))  # Logarithmic bucket distribution
```